### PR TITLE
RSDK-2506 Sessions client 

### PR DESCRIPTION
**Pull request by maximpertsov**
Implements client-side sessions for the python sdk

## TODO

* [x] Ensure in-progress actions are cancelled if the SDK is forcefully killed

## Manual Testing

* Copy this branch locally and `cd` into the repository.
* Copy the contents of https://gist.github.com/maximpertsov/acd5d9ddaaa835a000a310341cdfb1de into `testing.py` (or run the curl script below)

```
curl https://gist.githubusercontent.com/maximpertsov/acd5d9ddaaa835a000a310341cdfb1de/raw/0fe5f13d1cafef0eb39f7bfcb59556f7f21a53af/viam-python-sessions-testing.py > testing.py
```

* Find your favorite robot with a drive-able base and replace the `URI, SECRET, BASE_NAME` constants in script above with the relevant parameters.

* Run the following commands
```
poetry add -D aioconsole  # needed for the script above
make install && poetry run python testing.py
```

* You should see output that looks like this:
```
2023-05-25 17:21:09,768         DEBUG   viam.rpc.dial (dial.py:188)     Creating new viam-rust-utils runtime
2023-05-25 17:21:09,770         DEBUG   viam.rpc.dial (dial.py:209)     Dialing <URI> using viam-rust-utils library
2023-05-25 17:21:11,492         INFO    viam.rpc.dial (dial.py:239)     Connecting to socket: <SOCKET>
2023-05-25 17:21:11,624         DEBUG   viam.sessions_client (sessions_client.py:128)   sent heartbeat successfully
Move in a direction (WASD) or Q to exit
2023-05-25 17:21:12,038         DEBUG   viam.sessions_client (sessions_client.py:128)   sent heartbeat successfully
2023-05-25 17:21:12,450         DEBUG   viam.sessions_client (sessions_client.py:128)   sent heartbeat successfully
2023-05-25 17:21:12,865         DEBUG   viam.sessions_client (sessions_client.py:128)   sent heartbeat successfully
2023-05-25 17:21:13,279         DEBUG   viam.sessions_client (sessions_client.py:128)   sent heartbeat successfully
...
```

* Pressing `W A S or D + ENTER` should drive the base without any noticeable lag.

### CODE REVIEW

**Comment by edaniels** [comment]
why isn't this being implemented in rusttuils?

**Comment by maximpertsov** [comment]
> why isn't this being implemented in [rust-utils](https://github.com/viamrobotics/rust-utils/)?

the python sdk can still [dial a robot directly without using rust-utils](https://github.com/viamrobotics/viam-python-sdk/blob/7ae6a3cbc4f80295c61a4104a751a35d0d94836a/src/viam/rpc/dial.py#L233-L235), so we need to implement it here to support sessions for that kind of connection.

**Comment by edaniels** [comment]
> > why isn't this being implemented in [rust-utils](https://github.com/viamrobotics/rust-utils/)?
> 
> the python sdk can still [dial a robot directly without using rust-utils](https://github.com/viamrobotics/viam-python-sdk/blob/7ae6a3cbc4f80295c61a4104a751a35d0d94836a/src/viam/rpc/dial.py#L233-L235), so we need to implement it here to support sessions for that kind of connection.

got it

File: src/viam/sessions_client.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+            return
+
+        LOGGER.debug("resetting session")
+        self._supported = None
+
+    async def _send_request(self, event: SendRequest):
+        if self._disabled:
+            return
+
+        if event.method_name in [self.client.StartSession.name]:
```
Comment (edaniels) [inline src/viam/sessions_client.py:54]: do we need to exempt the other methods at https://github.com/viamrobotics/rdk/blob/main/session/doc.go#L71?

**Review by edaniels — APPROVED** [review]
LGTM assuming testing was done that the robot stopped the base once the python SDK was forcefully killed without warning

File: src/viam/sessions_client.py (hunk 1 above)
Comment (maximpertsov) [reply src/viam/sessions_client.py:54]: ah didn't see this before, will add

**Comment by maximpertsov** [comment]
> LGTM assuming testing was done that the robot stopped the base once the python SDK was forcefully killed without warning

Yeah, let me try this with a longer-running command to confirm

File: src/viam/sessions_client.py (hunk 1 above)
Comment (maximpertsov) [reply src/viam/sessions_client.py]: https://github.com/viamrobotics/viam-python-sdk/pull/301/commits/9f0c4e0f80406612d83d205c4d6df3e9da985fb2

File: src/viam/sessions_client.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        "/proto.rpc.v1.AuthService/Authenticate",
+        "/viam.robot.v1.RobotService/ResourceNames",
+        "/viam.robot.v1.RobotService/ResourceRPCSubtypes",
+        "/viam.robot.v1.RobotService/StartSession",
+        "/viam.robot.v1.RobotService/SendSessionHeartbeat",
+    ]
+)
+
+
+async def delay(coro, seconds):
```
Comment (njooma) [inline src/viam/sessions_client.py:32]: there's a `loop.call_later` function that might be useful here:
```python3
loop = asyncio.get_running_loop()
loop.call_later(seconds, coro)
```

File: src/viam/sessions_client.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        self._disabled = disabled
+
+        listen(self.channel, SendRequest, self._send_request)
+        listen(self.channel, RecvTrailingMetadata, self._recv_trailers)
+
+    def reset(self):
+        if self._lock.locked():
+            return
+
+        LOGGER.debug("resetting session")
```
Comment (njooma) [inline src/viam/sessions_client.py:61]: [nit] can we make all logs `Sentence case` please?

File: src/viam/sessions_client.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+                response = await self.client.StartSession(request)
+            except GRPCError as error:
+                if error.status == Status.UNIMPLEMENTED:
+                    self._supported = False
+                    return self._metadata
+                else:
+                    raise
+            else:
+                if response is None:
+                    raise GRPCError(status=Status.INTERNAL, message="expected response to start session")
```
Comment (njooma) [inline src/viam/sessions_client.py:103]: Same with error messages as with logs: `Sentence case` please

File: pyproject.toml [hunk 5]
```diff
@@ -35,6 +35,7 @@ types-pillow = "^9.5.0.4"
 mypy-protobuf = "^3.4.0"
 tox = "^4.5.1"
 isort = "^5.12.0"
+pytest-watcher = "^0.2.6"
```
Comment (njooma) [inline pyproject.toml:38]: [q] is this needed?

**Review by njooma — APPROVED** [review]
Just looked over for code quality, didn't actually test.

Looks good! Mostly nits

File: pyproject.toml (hunk 5 above)
Comment (maximpertsov) [reply pyproject.toml:38]: Not strictly, but it was helpful to have this utility when I was spiking out unit tests. I can you remove if you prefer.

File: src/viam/sessions_client.py (hunk 2 above)
Comment (maximpertsov) [reply src/viam/sessions_client.py:32]: This looked promising, but unfortunately `loop.call_later` takes a sync callback and not a co-routine: https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.call_later

File: pyproject.toml (hunk 5 above)
Comment (maximpertsov) [reply pyproject.toml]: Went ahead and removed

**Pull request merged**